WalletWasabi/Userfacing/Bip21/Bip21UriParser.cs

```diff
@@ -0,0 +1,178 @@
+	public static readonly Error ErrorInvalidAmountValue = new(Code: 8, "Invalid amount value.");
+	public static readonly Error ErrorUnsupportedReqParameter = new(Code: 9, "Unsupported required parameter found.");
+
+	public static bool TryParse(string input, Network network, [NotNullWhen(true)] out Result? result, [NotNullWhen(false)] out Error? error)
+	{
+		result = null;
+		error = null;
+
+		if (!Uri.TryCreate(input, UriKind.Absolute, out Uri? parsedUri))
+		{
+			error = ErrorInvalidUri with { Details = input };
+			return false;
+		}
+
+		if (!parsedUri.Scheme.Equals(UriScheme, StringComparison.OrdinalIgnoreCase))
+		{
+			error = ErrorInvalidUriScheme with { Details = parsedUri.Scheme };
+			return false;
+		}
+
+		if (parsedUri.AbsolutePath is not { Length: > 0 } addressString)
+		{
+			error = ErrorMissingAddress;
+			return false;
+		}
+
+		Money? amount = null;
+		string? label = null;
+		string? message = null;
+
+		if (!TryParseBitcoinAddressForNetwork(addressString, network, out BitcoinAddress? address))
+		{
+			error = ErrorInvalidAddress with { Details = addressString };
+			return false;
+		}
+
+		Dictionary<string, string> unknownParameters = new();
+		NameValueCollection queryParameters = HttpUtility.ParseQueryString(parsedUri.Query);
+
+		foreach (string? parameterName in queryParameters.AllKeys)
+		{
+			if (parameterName is null)
+			{
+				continue;
+			}
+
+			string? value = queryParameters[parameterName];
+
+			if (value is null)
+			{
+				continue;
+			}
+
+			if (parameterName == "amount")
+			{
+				if (amount is not null)
+				{
+					error = ErrorDuplicateParameter with { Details = parameterName };
+					return false;
+				}
+
+				if (value.Trim() == "")
+				{					
+					error = ErrorMissingAmountValue with { Details = value };
+					return false;
+				}
+
+				if (!Money.TryParse(value, out amount))
+				{
+					error = ErrorInvalidAmountValue with { Details = value };
+					return false;
+				}
+			}
+			else if (parameterName == "label")
+			{
+				if (label is not null)
+				{
+					error = ErrorDuplicateParameter with { Details = parameterName };
+					return false;
+				}
+
+				label = value;
+			}
+			else if (parameterName == "message")
+			{
+				if (message is not null)
+				{
+					error = ErrorDuplicateParameter with { Details = parameterName };
+					return false;
+				}
+
+				message = value;
+			}
+			else if (parameterName.StartsWith("req-", StringComparison.Ordinal))
+			{
+				error = ErrorUnsupportedReqParameter with { Details = parameterName };
+				return false;
+			}
+		}
+
+		result = new(Uri: parsedUri, network, address, amount, Label: label, Message: message, unknownParameters);
+		return true;
+	}
+
+	/// <remarks>NBitcoin does not provide a try-parse method.</remarks>
+	public static bool TryParseBitcoinAddressForNetwork(string address, Network network, [NotNullWhen(true)] out BitcoinAddress? bitcoinAddress)
```

Comment:
This is not BIP21. It is confusing to have this here under this class, isn't it?

Reply:
I did not understand why you have two TryParse methods and one more in AddressStringParser. Later I realized that AddressStringParser.TryParse is general and you created a class only for BIP21. In that sense, TryParseBitcoinAddressForNetwork is mixing the abstractions.

Reply:
Beautiful.